### PR TITLE
Fix attachment download bug

### LINKED ISSUE #27204: [BUG][API] Issue/pr attachments return wrong download url
### Description

The issue attachment API return the following:

```
[
	{
		"id": 58540,
		"name": "Screenshot_20230919_133134.png",
		"size": 41611,
		"download_count": 16,
		"created_at": "2023-09-19T08:37:33Z",
		"uuid": "c9bf83dc-8d08-460e-9986-ee7f292bccd7",
		"browser_download_url": "https://codeberg.org/api/repos/gitnex/GitNex/releases/0/assets/58540"
	},
	{
		"id": 58541,
		"name": "Screenshot_20230919_133201.png",
		"size": 41475,
		"download_count": 17,
		"created_at": "2023-09-19T08:37:33Z",
		"uuid": "e5f07ce2-4179-43ae-81a2-09e4d319aab2",
		"browser_download_url": "https://codeberg.org/api/repos/gitnex/GitNex/releases/0/assets/58541"
	},
	{
		"id": 58542,
		"name": "Screenshot_20230919_133224.png",
		"size": 19559,
		"download_count": 16,
		"created_at": "2023-09-19T08:37:34Z",
		"uuid": "6bed5cd4-2f82-4566-bea7-605d12af498d",
		"browser_download_url": "https://codeberg.org/api/repos/gitnex/GitNex/releases/0/assets/58542"
	}
]
```

This is a possible bug because the `browser_download_url` is broken to me. This may also affect the comments attachment API too.
The `browser_download_url` should be a proper downloadable URL like how it is in the releases.

### Gitea Version

1.20.3

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Codeberg and Gitea 1.20.3 for testing on local setup.

https://codeberg.org/api/v1/repos/gitnex/GitNex/issues/1300/assets

### Database

None

### PR DESCRIPTION
**Pull request by lunny**
Fix #27204

This PR allows `/<username>/<reponame>/attachments/<uuid>` access with personal access token and also changed attachments API download url to it so it can be download correctly.

### CODE REVIEW

File: services/auth/oauth2.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (o *OAuth2) userIDFromToken(ctx context.Context, tokenSHA string, store Dat
 func (o *OAuth2) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error) {
 	// These paths are not API paths, but we still want to check for tokens because they maybe in the API returned URLs
 	if !middleware.IsAPIPath(req) && !isAttachmentDownload(req) && !isAuthenticatedTokenRequest(req) &&
-		!gitRawReleasePathRe.MatchString(req.URL.Path) {
+		!gitRawOrAttachPathRe.MatchString(req.URL.Path) {
```
Comment (delvh) [inline services/auth/oauth2.go]: I think that should be the method instead of a re-implementation.

File: services/convert/attachment.go [hunk 2]
```diff
@@ -16,12 +13,7 @@ func WebAssetDownloadURL(repo *repo_model.Repository, attach *repo_model.Attachm
 }
 
 func APIAssetDownloadURL(repo *repo_model.Repository, attach *repo_model.Attachment) string {
```
Comment (delvh) [inline services/convert/attachment.go:15]: Do we even need the `repo` param anymore now?

Comment (lunny) [reply services/convert/attachment.go:15]: I think these function could be removed. To keep this PR small, I need another refactor PR to do that.

File: services/auth/oauth2.go (hunk 1 above)
Comment (lunny) [reply services/auth/oauth2.go]: Resolved.

**Review by delvh — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

File: services/convert/attachment.go [hunk 3]
```diff
@@ -16,12 +13,7 @@ func WebAssetDownloadURL(repo *repo_model.Repository, attach *repo_model.Attachm
...
-	if attach.CustomDownloadURL != "" {
-		return attach.CustomDownloadURL
-	}
-
-	// /repos/{owner}/{repo}/releases/{id}/assets/{attachment_id}
-	return setting.AppURL + "api/repos/" + repo.FullName() + "/releases/" + strconv.FormatInt(attach.ReleaseID, 10) + "/assets/" + strconv.FormatInt(attach.ID, 10)
```
Comment (wxiaoguang) [inline services/convert/attachment.go:24]: What would happen to `setting.AppURL + "api/repos/" + repo.FullName() + "/releases/" + strconv.FormatInt(attach.ReleaseID, 10) + "/assets/" + strconv.FormatInt(attach.ID, 10)` ?

Should this handler be removed?

Comment (lunny) [reply services/convert/attachment.go:24]: Since it has been released, maybe deprecated is better.

**Pull request merged**